specification/dartLangSpec.tex

```diff
@@ -1317,127 +1342,240 @@ \section{Variables}
 A variable declared at the top-level of a library is referred to as either a
 \IndexCustom{library variable}{variable!library} or a
 \IndexCustom{top-level variable}{variable!top-level}.
+It is a \Error{compile-time error} if a library variable declaration
+has the modifier \STATIC.
+
+\LMHash{}%
+A \IndexCustom{class variable}{variable!class}
+is a variable whose declaration is
+immediately nested inside a class declaration and
+includes the modifier \STATIC.
 
 \LMHash{}%
 A \IndexCustom{static variable}{variable!static}
 is a variable that is not associated with a particular instance,
 but rather with an entire library or class.
 Static variables include library variables and class variables.
-Class variables are variables whose declaration is
-immediately nested inside a class declaration and
-includes the modifier \STATIC.
-A library variable is implicitly static.
-It is a compile-time error to preface a top-level variable declaration
-with the built-in identifier (\ref{identifierReference}) \STATIC.
+
+\LMHash{}%
+A \Error{compile-time error} occurs if a static variable has a
+no initializing expression and a type which is not nullable
+(\ref{typeNullability}),
+unless the variable declaration has
+the modifier \LATE{} or the modifier \EXTERNAL.
+
+\LMHash{}%
+A \IndexCustom{non-local variable}{variable!non-local}
+is a library variable, a class variable, or an instance variable.
+\commentary{%
+That is, any kind of variable which is not a local variable.%
+}
 
 \LMHash{}%
 A \IndexCustom{constant variable}{variable!constant}
 is a variable whose declaration includes the modifier \CONST.
 A constant variable must be initialized to a constant expression
-(\ref{constants})
-or a compile-time error occurs.
+(\ref{constants}),
+or a \Error{compile-time error} occurs.
 
 \commentary{%
-An initializing expression of a constant variable occurs in a constant context
-(\ref{constantContexts}),
-which means that \CONST{} modifiers need not be specified explicitly.%
+An initializing expression $e$ of a constant variable declaration
+occurs in a constant context
+(\ref{constantContexts}).
+This means that \CONST{} modifiers in $e$ need not be specified explicitly.
+
+It is grammatically impossible for a constant variable declaration
+to have the modifier \LATE.
+However, even if it had been grammatically possible,
+a \LATE{} constant variable would still have to be a compile-time error,
+because being a compile-time constant is inherently incompatible
+with being computed late.%
 }
 
 \LMHash{}%
-A \IndexCustom{final variable}{variable!final}
+An \IndexCustom{immutable variable}{variable!immutable}
 is a variable whose binding is fixed upon initialization;
```

Comment:
Have we defined "initialization".

We use "bound to a value" below for most cases.
Does it cover both local and non-local variables?